In [102]:
import numpy as np
import pandas as pd

In [103]:
dataset = pd.read_csv('BostonHousing.csv')
dataset

crim    zn  indus  chas    nox     rm   age     dis  rad  tax  \
0    0.00632  18.0   2.31     0  0.538  6.575  65.2  4.0900    1  296   
1    0.02731   0.0   7.07     0  0.469  6.421  78.9  4.9671    2  242   
2    0.02729   0.0   7.07     0  0.469  7.185  61.1  4.9671    2  242   
3    0.03237   0.0   2.18     0  0.458  6.998  45.8  6.0622    3  222   
4    0.06905   0.0   2.18     0  0.458  7.147  54.2  6.0622    3  222   
..       ...   ...    ...   ...    ...    ...   ...     ...  ...  ...   
501  0.06263   0.0  11.93     0  0.573  6.593  69.1  2.4786    1  273   
502  0.04527   0.0  11.93     0  0.573  6.120  76.7  2.2875    1  273   
503  0.06076   0.0  11.93     0  0.573  6.976  91.0  2.1675    1  273   
504  0.10959   0.0  11.93     0  0.573  6.794  89.3  2.3889    1  273   
505  0.04741   0.0  11.93     0  0.573  6.030  80.8  2.5050    1  273   

     ptratio       b  lstat  medv  
0       15.3  396.90   4.98  24.0  
1       17.8  396.90   9.14  21.6  
2       17.8  392.83   4.03  34.7  
3       18.7  394.63   2.94  33.4  
4       18.7  396.90   5.33  36.2  
..       ...     ...    ...   ...  
501     21.0  391.99   9.67  22.4  
502     21.0  396.90   9.08  20.6  
503     21.0  396.90   5.64  23.9  
504     21.0  393.45   6.48  22.0  
505     21.0  396.90   7.88  11.9  

[506 rows x 14 columns]

In [104]:
x = dataset.iloc[:,:-1]
y = dataset.iloc[:,-1]

In [105]:
x

crim    zn  indus  chas    nox     rm   age     dis  rad  tax  \
0    0.00632  18.0   2.31     0  0.538  6.575  65.2  4.0900    1  296   
1    0.02731   0.0   7.07     0  0.469  6.421  78.9  4.9671    2  242   
2    0.02729   0.0   7.07     0  0.469  7.185  61.1  4.9671    2  242   
3    0.03237   0.0   2.18     0  0.458  6.998  45.8  6.0622    3  222   
4    0.06905   0.0   2.18     0  0.458  7.147  54.2  6.0622    3  222   
..       ...   ...    ...   ...    ...    ...   ...     ...  ...  ...   
501  0.06263   0.0  11.93     0  0.573  6.593  69.1  2.4786    1  273   
502  0.04527   0.0  11.93     0  0.573  6.120  76.7  2.2875    1  273   
503  0.06076   0.0  11.93     0  0.573  6.976  91.0  2.1675    1  273   
504  0.10959   0.0  11.93     0  0.573  6.794  89.3  2.3889    1  273   
505  0.04741   0.0  11.93     0  0.573  6.030  80.8  2.5050    1  273   

     ptratio       b  lstat  
0       15.3  396.90   4.98  
1       17.8  396.90   9.14  
2       17.8  392.83   4.03  
3       18.7  394.63   2.94  
4       18.7  396.90   5.33  
..       ...     ...    ...  
501     21.0  391.99   9.67  
502     21.0  396.90   9.08  
503     21.0  396.90   5.64  
504     21.0  393.45   6.48  
505     21.0  396.90   7.88  

[506 rows x 13 columns]

In [106]:
y

0      24.0
1      21.6
2      34.7
3      33.4
4      36.2
       ... 
501    22.4
502    20.6
503    23.9
504    22.0
505    11.9
Name: medv, Length: 506, dtype: float64

In [107]:
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test = train_test_split(x,y,test_size = 0.2,random_state = 5)

In [108]:
x_train.shape

(404, 13)

In [109]:
x_test.shape

(102, 13)

In [110]:
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
x_train = scaler.fit_transform(x_train)
x_test = scaler.transform(x_test)

In [111]:
print(x_train)

[[-2.76469903e-01 -5.07766334e-01 -4.02011945e-01 ...  1.17561917e+00
  -3.01968083e-04  8.35287872e-01]
 [-4.07463681e-01  3.33136528e+00 -1.43188822e+00 ... -2.20285923e+00
   4.12907260e-01 -1.32549534e+00]
 [ 8.06483214e+00 -5.07766334e-01  1.07815655e+00 ...  8.10378266e-01
  -3.84927565e+00  1.15819623e+00]
 ...
 [-3.88566725e-01 -5.07766334e-01 -1.71664438e-01 ...  3.53827130e-01
   3.87833612e-01  7.28600090e-01]
 [-3.82724087e-01 -5.07766334e-01 -3.79142727e-02 ...  7.98964487e-02
   4.05598798e-01 -2.14519905e-01]
 [-3.96910057e-01  2.90479510e+00 -1.32786031e+00 ...  1.63217031e+00
   1.93878260e-01 -9.82671938e-01]]


In [112]:
print(x_test)

[[-0.36515584 -0.50776633 -0.69031786 ... -0.46796491  0.32138282
  -1.32976286]
 [-0.40502752  2.9047951  -0.87608198 ...  0.35382713  0.42842369
  -1.10642977]
 [-0.40379561 -0.50776633 -1.10494337 ... -0.28534446  0.37535301
  -0.52178072]
 ...
 [-0.12723949 -0.50776633  1.29810127 ... -1.70065298 -3.037699
   0.52091454]
 [ 0.26196199 -0.50776633  1.07815655 ...  0.81037827  0.42842369
  -0.24723749]
 [-0.39024922  0.43068806 -0.7408457  ...  0.30817202  0.17892402
  -0.42078295]]


In [113]:
from sklearn.linear_model import LinearRegression
model = LinearRegression()
model.fit(x_train,y_train)

LinearRegression()

In [114]:
y_pred = model.predict(x_test)
y_pred

array([37.56311787, 32.14445143, 27.06573629,  5.67080633, 35.09982577,
        5.85803701, 27.53708506, 31.81019188, 26.35634771, 22.77208748,
       31.91183048, 21.50224061, 23.70119983, 33.3622504 , 28.51633591,
       14.39456899,  0.19284025, 18.66247155, 13.71004139, 14.13408635,
        2.03263952, 19.7280831 , 38.18657429, 24.19760058, 31.30247973,
       11.14144544, 25.03636951, 23.27970871, 22.49420127, 20.52972594,
       15.16513744,  6.92553586, 18.3557733 , 22.37179804, 28.91287973,
       19.02980786, 30.19357214,  8.74384915, 40.86691522, 34.53763591,
       20.70224878,  2.59618963, 29.99590282, 12.15704798, 27.10186397,
       30.8052437 , -6.24169079, 19.84885777, 20.92973441, 12.43523958,
       20.4949947 , 19.19231742, 23.69073157, 12.67998473, 17.14252424,
       25.04649176, 34.77758126, 15.23294903, 28.22306193, 21.08745388,
       20.39506129, 25.79476888, 14.72463673, 33.18635032, 23.17771307,
       13.11057248, 19.23154617, 24.61162961, 21.50327036, 22.00

In [115]:
from sklearn.metrics import r2_score
from sklearn.model_selection import cross_val_score
print(np.mean(cross_val_score(model,x_train,y_train,cv = 5, scoring = 'r2')))
print(np.mean(cross_val_score(model,x_test,y_test,cv = 5, scoring = 'r2')))

0.701878817261314
0.65392094832943


#### We can clearly see that there is a large gap in explainablity of variance in training and testing data. so, we can say that there is an overfitting in model.

In [116]:
from mlxtend.feature_selection import ExhaustiveFeatureSelector
EFS = ExhaustiveFeatureSelector(model,max_features = 13, scoring = 'r2',cv = 10, n_jobs = -1)
sel = EFS.fit(x_train,y_train)

Features: 8191/8191

In [117]:
sel.best_feature_names_

('0', '1', '4', '5', '7', '8', '9', '10', '11', '12')

#### It is giving numbers as, the column names were removed as we performed the standardization and dataframe was converted into numpy array.

In [118]:
sel.best_score_

0.7042709681532113

In [119]:
x_sel = sel.subsets_
x_sel = pd.DataFrame(x_sel).T
x_sel.sort_values('avg_score',ascending=False)
# x_sel

feature_idx  \
7975        (0, 1, 4, 5, 7, 8, 9, 10, 11, 12)   
8141     (0, 1, 2, 4, 5, 7, 8, 9, 10, 11, 12)   
8150     (0, 1, 3, 4, 5, 7, 8, 9, 10, 11, 12)   
8153     (0, 1, 4, 5, 6, 7, 8, 9, 10, 11, 12)   
8183  (0, 1, 2, 3, 4, 5, 7, 8, 9, 10, 11, 12)   
...                                       ...   
53                                    (3, 11)   
11                                      (11,)   
49                                     (3, 7)   
7                                        (7,)   
3                                        (3,)   

                                              cv_scores avg_score  \
7975  [0.6976271527384033, 0.5893677473731809, 0.756...  0.704271   
8141  [0.6977135685685397, 0.5893451722940015, 0.756...  0.703106   
8150  [0.7061355032371863, 0.6113833891208393, 0.735...  0.701067   
8153  [0.6964838217213023, 0.5858358077270205, 0.756...  0.700756   
8183  [0.7058578946614813, 0.6111956117116959, 0.734...  0.700082   
...                                                 ...       ...   
53    [0.15069892023120168, 0.26121009220585834, -0....  0.088098   
11    [0.07226066876710824, 0.1576198930480317, -0.0...   0.08497   
49    [0.1510058508041685, 0.209778329701358, -0.054...   0.05262   
7     [0.05059318867330542, 0.04905966623831215, -0....  0.036324   
3     [0.07616229801951102, 0.10131740452235205, 0.0... -0.005653   

                                feature_names  
7975        (0, 1, 4, 5, 7, 8, 9, 10, 11, 12)  
8141     (0, 1, 2, 4, 5, 7, 8, 9, 10, 11, 12)  
8150     (0, 1, 3, 4, 5, 7, 8, 9, 10, 11, 12)  
8153     (0, 1, 4, 5, 6, 7, 8, 9, 10, 11, 12)  
8183  (0, 1, 2, 3, 4, 5, 7, 8, 9, 10, 11, 12)  
...                                       ...  
53                                    (3, 11)  
11                                      (11,)  
49                                     (3, 7)  
7                                        (7,)  
3                                        (3,)  

[8191 rows x 4 columns]

In [120]:
def AdjR2(r2,n,k):
    return 1 - ((1-r2)*(n-1)/(n-k-1))

In [121]:
x_sel['observations'] = 404
x_sel['no. of features'] = x_sel['feature_idx'].apply(lambda x: len(x))
x_sel['Adjusted R2'] = AdjR2(x_sel['avg_score'],x_sel.shape[0],x_sel['no. of features'])
x_sel.sort_values('Adjusted R2',ascending = False)
# x_sel

feature_idx  \
7975        (0, 1, 4, 5, 7, 8, 9, 10, 11, 12)   
8141     (0, 1, 2, 4, 5, 7, 8, 9, 10, 11, 12)   
8150     (0, 1, 3, 4, 5, 7, 8, 9, 10, 11, 12)   
8153     (0, 1, 4, 5, 6, 7, 8, 9, 10, 11, 12)   
8183  (0, 1, 2, 3, 4, 5, 7, 8, 9, 10, 11, 12)   
...                                       ...   
53                                    (3, 11)   
11                                      (11,)   
49                                     (3, 7)   
7                                        (7,)   
3                                        (3,)   

                                              cv_scores avg_score  \
7975  [0.6976271527384033, 0.5893677473731809, 0.756...  0.704271   
8141  [0.6977135685685397, 0.5893451722940015, 0.756...  0.703106   
8150  [0.7061355032371863, 0.6113833891208393, 0.735...  0.701067   
8153  [0.6964838217213023, 0.5858358077270205, 0.756...  0.700756   
8183  [0.7058578946614813, 0.6111956117116959, 0.734...  0.700082   
...                                                 ...       ...   
53    [0.15069892023120168, 0.26121009220585834, -0....  0.088098   
11    [0.07226066876710824, 0.1576198930480317, -0.0...   0.08497   
49    [0.1510058508041685, 0.209778329701358, -0.054...   0.05262   
7     [0.05059318867330542, 0.04905966623831215, -0....  0.036324   
3     [0.07616229801951102, 0.10131740452235205, 0.0... -0.005653   

                                feature_names  observations  no. of features  \
7975        (0, 1, 4, 5, 7, 8, 9, 10, 11, 12)           404               10   
8141     (0, 1, 2, 4, 5, 7, 8, 9, 10, 11, 12)           404               11   
8150     (0, 1, 3, 4, 5, 7, 8, 9, 10, 11, 12)           404               11   
8153     (0, 1, 4, 5, 6, 7, 8, 9, 10, 11, 12)           404               11   
8183  (0, 1, 2, 3, 4, 5, 7, 8, 9, 10, 11, 12)           404               12   
...                                       ...           ...              ...   
53                                    (3, 11)           404                2   
11                                      (11,)           404                1   
49                                     (3, 7)           404                2   
7                                        (7,)           404                1   
3                                        (3,)           404                1   

     Adjusted R2  
7975    0.703909  
8141    0.702707  
8150    0.700665  
8153    0.700353  
8183    0.699642  
...          ...  
53      0.087875  
11      0.084858  
49      0.052388  
7       0.036207  
3      -0.005776  

[8191 rows x 7 columns]

#### Here, the first row has the maximum Adusted R2 score which means that the set of features (0, 1, 4, 5, 7, 8, 9, 10, 11, 12) will be able to explain the variability in dependent variable best.

In [122]:
x_train = sel.transform(x_train)
x_train = pd.DataFrame(x_train)

In [123]:
x_train

0         1         2         3         4         5         6  \
0   -0.276470 -0.507766 -0.118484 -0.848834 -0.022455 -0.636453 -0.583618   
1   -0.407464  3.331365 -1.314148  2.361809  0.999022 -0.981745 -1.231926   
2    8.064832 -0.507766  1.112090 -0.478931 -0.988813  1.665490  1.551633   
3   -0.403977 -0.507766  0.186977 -0.243407 -0.752702 -0.981745 -0.785842   
4    0.696610 -0.507766  1.234274  0.166954 -1.068474  1.665490  1.551633   
..        ...       ...       ...       ...       ...       ...       ...   
399  0.246720 -0.507766  0.274252 -0.252077 -0.139951  1.665490  1.551633   
400 -0.399548  1.411799 -0.999959  1.295448  0.357105 -0.521356 -0.042371   
401 -0.388567 -0.507766  0.291707 -0.893627 -0.503833 -0.406259 -0.084005   
402 -0.382724 -0.507766 -0.546130  0.054249  0.253974 -0.636453 -0.762051   
403 -0.396910  2.904795 -1.209418 -0.509275  3.287925 -0.636453 -0.423028   

            7         8         9  
0    1.175619 -0.000302  0.835288  
1   -2.202859  0.412907 -1.325495  
2    0.810378 -3.849276  1.158196  
3    1.175619  0.428424 -0.483373  
4    0.810378  0.194665  1.114099  
..        ...       ...       ...  
399  0.810378  0.380975  0.355904  
400 -1.472377  0.428424 -1.008277  
401  0.353827  0.387834  0.728600  
402  0.079896  0.405599 -0.214520  
403  1.632170  0.193878 -0.982672  

[404 rows x 10 columns]

In [124]:
x_test = sel.transform(x_test)
x_test = pd.DataFrame(x_test)
x_test

0         1         2         3         4         5         6  \
0   -0.365156 -0.507766 -0.415218  2.530867 -0.300735 -0.176064 -0.583618   
1   -0.405028  2.904795 -1.226873  0.493509  0.624915 -0.636453 -0.952380   
2   -0.403796 -0.507766 -0.546130  0.185738 -0.360822 -0.866648 -0.803686   
3    0.870897 -0.507766  1.016087 -1.997557 -1.294799  1.665490  1.551633   
4   -0.383624  0.345374 -0.764317  2.026585  0.259622 -0.751550 -1.083231   
..        ...       ...       ...       ...       ...       ...       ...   
97  -0.405853  1.198514 -1.069779  0.939994  2.414571 -0.981745 -0.417080   
98  -0.384165 -0.507766 -0.546130 -0.695671 -0.087996 -0.636453 -0.762051   
99  -0.127239 -0.507766  2.787765 -1.468711 -1.021048 -0.521356 -0.012632   
100  0.261962 -0.507766 -0.170849 -0.067125 -0.199210  1.665490  1.551633   
101 -0.390249  0.430688 -1.052324  0.208857  1.944393 -0.291162 -0.446819   

            7         8         9  
0   -0.467965  0.321383 -1.329763  
1    0.353827  0.428424 -1.106430  
2   -0.285344  0.375353 -0.521781  
3    0.810378  0.428424  3.171039  
4    0.079896  0.359499 -0.838999  
..        ...       ...       ...  
97   0.582103  0.349155 -0.937152  
98   0.079896  0.361411  0.505267  
99  -1.700653 -3.037699  0.520915  
100  0.810378  0.428424 -0.247237  
101  0.308172  0.178924 -0.420783  

[102 rows x 10 columns]

In [125]:
model.fit(x_train,y_train)

LinearRegression()

In [126]:
y_pred_2 = model.predict(x_test)
y_pred

array([37.56311787, 32.14445143, 27.06573629,  5.67080633, 35.09982577,
        5.85803701, 27.53708506, 31.81019188, 26.35634771, 22.77208748,
       31.91183048, 21.50224061, 23.70119983, 33.3622504 , 28.51633591,
       14.39456899,  0.19284025, 18.66247155, 13.71004139, 14.13408635,
        2.03263952, 19.7280831 , 38.18657429, 24.19760058, 31.30247973,
       11.14144544, 25.03636951, 23.27970871, 22.49420127, 20.52972594,
       15.16513744,  6.92553586, 18.3557733 , 22.37179804, 28.91287973,
       19.02980786, 30.19357214,  8.74384915, 40.86691522, 34.53763591,
       20.70224878,  2.59618963, 29.99590282, 12.15704798, 27.10186397,
       30.8052437 , -6.24169079, 19.84885777, 20.92973441, 12.43523958,
       20.4949947 , 19.19231742, 23.69073157, 12.67998473, 17.14252424,
       25.04649176, 34.77758126, 15.23294903, 28.22306193, 21.08745388,
       20.39506129, 25.79476888, 14.72463673, 33.18635032, 23.17771307,
       13.11057248, 19.23154617, 24.61162961, 21.50327036, 22.00

In [127]:
print("Training:",np.mean(cross_val_score(model,x_train,y_train,cv = 5 , scoring= 'r2')))
print("Test:",np.mean(cross_val_score(model,x_test,y_test,cv = 5 , scoring= 'r2')))


Training: 0.707911975474306
Test: 0.6910480978849218


### Prior to the feature selection based on the "EXHAUSTIVE FEATURE SELECTION", the R2 Score for training data was 0.701878817261314 and for test data it was 0.65392094832943.

### But after the selection of features the R2 socre for training data was 0.707911975474306 and for Test it was : 0.6910480978849218. So we can say the overfitting which was happening has been reduced by the feature selection.